plugins/rebalance/rest.go

```diff
@@ -0,0 +1,323 @@
+		return
+	}
+
+	// Check rebalance process is started or not
+	if rebalinfo.Status != rebalanceapi.Started {
+		restutils.SendHTTPError(r.Context(), w, http.StatusBadRequest, errors.ErrRebalanceNotStarted)
+		return
+	}
+
+	//A simple transaction to stop rebalance
+	txn := transaction.NewTxn(ctx)
+	defer txn.Cleanup()
+	lock, unlock, err := transaction.CreateLockSteps(volname)
+	if err != nil {
+		restutils.SendHTTPError(r.Context(), w, http.StatusInternalServerError, err.Error())
+		return
+	}
+
+	txn.Nodes = vol.Nodes()
+	txn.Steps = []*transaction.Step{
+		lock,
+		{
+			DoFunc: "rebalance-stop",
+			Nodes:  txn.Nodes,
+		},
+		{
+			DoFunc: "rebalance-store",
+			Nodes:  []uuid.UUID{gdctx.MyUUID},
+		},
+		unlock,
+	}
+
+	err = txn.Ctx.Set("volname", volname)
+	if err != nil {
+		logger.WithError(err).Error("failed to set volname in transaction context")
+		restutils.SendHTTPError(ctx, w, http.StatusInternalServerError, err.Error())
+		return
+	}
+
+	rebalinfo.Volname = volname
+	rebalinfo.Status = rebalanceapi.Stopped
+	rebalinfo.Cmd = rebalanceapi.CmdStop
+
+	err = txn.Ctx.Set("rinfo", rebalinfo)
+	if err != nil {
+		logger.WithError(err).Error("failed to set rebalance info in transaction context")
+		restutils.SendHTTPError(ctx, w, http.StatusInternalServerError, err.Error())
+		return
+	}
+
+	err = txn.Do()
+	if err != nil {
+		logger.WithFields(log.Fields{
+			"error":   err.Error(),
+			"volname": volname,
+		}).Error("failed to stop rebalance on volume")
+		restutils.SendHTTPError(r.Context(), w, http.StatusInternalServerError, err.Error())
+		return
+	}
+
+	txn.Ctx.Logger().WithField("volname", rebalinfo.Volname).Info("rebalance stopped")
+	restutils.SendHTTPResponse(r.Context(), w, http.StatusOK, rebalinfo)
+}
+
+
+
+
+func rebalanceStatusHandler(w http.ResponseWriter, r *http.Request) {
+	ctx := r.Context()
+
+	// collect inputs from url
+	volname := mux.Vars(r)["volname"]
+
+	// Validate rebalance command
+	vol, err := volume.GetVolume(volname)
+	if err != nil {
+		restutils.SendHTTPError(ctx, w, http.StatusNotFound, errors.ErrVolNotFound)
+		return
+	}
+
+	rebalinfo, err := GetRebalanceInfo(volname)
+	if err != nil {
+		restutils.SendHTTPError(ctx, w, http.StatusNotFound, err.Error())
+		return
+	}
+
+
+        // Get the consolidated status from all the nodes
+
+	txn := transaction.NewTxn(ctx)
+	defer txn.Cleanup()
+	lock, unlock, err := transaction.CreateLockSteps(volname)
+	if err != nil {
+		restutils.SendHTTPError(r.Context(), w, http.StatusInternalServerError, err.Error())
+		return
+	}
+
+	err = txn.Ctx.Set("volname", volname)
+	if err != nil {
+		txn.Ctx.Logger().WithError(err).Error("failed to set volname in transaction context")
+		restutils.SendHTTPError(ctx, w, http.StatusInternalServerError, err.Error())
+		return
+	}
+
+	txn.Nodes = vol.Nodes()
+	txn.Steps = []*transaction.Step{
+		lock,
+		{
+			DoFunc: "rebalance-status",
+			Nodes:  txn.Nodes,
+		},
+		{
+			DoFunc: "rebalance-store",
```

Comment:
I think we can skip this step. Every time when status is requested, we get status from running rebalance daemon(except complete), so we can skip storing in etcd.

Reply:
We would need to get the status from any completed rebalance processes as well and those will be stored in etcd. The rebalance processes on different nodes may not finish at the same time. Do you mean we query etcd only if the process is not running?